### PR TITLE
Issue-2031 Fixed Grade Column Sorting and Saving

### PR DESCRIPTION
**Pull request by peterzhao333**

### CODE REVIEW

File: app/views/grade_entry_forms/_react_grades_table.js.jsx.erb [hunk 1]
```diff
@@ -176,17 +191,14 @@
             r[column_id] = <img src={row[column_id]}/>;
           } 
           else if (column_id == 'total_marks'){
-            r[column_id] = <input id={'grade_' + row.id + '_' + column.id} name={'grade_' + row.id + '_' + column.id} data-grade-entry-item-id={column.id} data-student-id={row.id}  type="text" size="4" value={row[column_id]} className="grade-input" disabled/>;
+            r[column_id] = <GradeBoxTotal grade_id={'grade_' + row.id + '_' + column.id} grade_entry_column={column.id} student_id={row.id} value={row[column_id]} disabled/>;
```
Comment (david-yz-liu) [inline app/views/grade_entry_forms/_react_grades_table.js.jsx.erb]: Could you please break these into multiple lines?

File: app/assets/javascripts/ReactComponents/table.js [hunk 2]
```diff
@@ -500,12 +500,24 @@ function sort_by_column(data, column, direction, compare) {
     if (typeof a == 'string') {
       return a.toLowerCase().replace(' ', '');
     } else if (a.hasOwnProperty('props')) {
+      // Is a react Grade Box
+      if (a.props.hasOwnProperty('grade_entry_column') || a.props.hasOwnProperty('data-grade-entry-item-id')){
```
Comment (david-yz-liu) [inline app/assets/javascripts/ReactComponents/table.js]: Line too long? Also, space before `{`

File: app/assets/javascripts/ReactComponents/table.js [hunk 3]
```diff
@@ -500,12 +500,24 @@ function sort_by_column(data, column, direction, compare) {
...
+        return a;
+      // Cotains image
+      } else if (a.props.hasOwnProperty('src')) {
+          return a.props.src;
```
Comment (david-yz-liu) [inline app/assets/javascripts/ReactComponents/table.js]: Indentation is 2 spaces, it appears.

File: app/assets/javascripts/ReactComponents/table.js [hunk 4]
```diff
@@ -500,12 +500,24 @@ function sort_by_column(data, column, direction, compare) {
...
-    } else {
-      return a;
+      } else {
+          return a.props.dangerouslySetInnerHTML.__html.toLowerCase();
+      }
     }
+      return a;
   }
+
+  if (typeof compare == 'string'){
```
Comment (david-yz-liu) [inline app/assets/javascripts/ReactComponents/table.js]: Use `===`

File: app/views/grade_entry_forms/_react_grades_table.js.jsx.erb [hunk 5]
```diff
@@ -16,24 +16,66 @@
...
     * checked from the Ruby controller
     */
   function update_cell(cell, value) {
-    // If the cell exists, change it's value to the one supplied
+    // If the total marks cell exists, change it's value to the one supplied
     if (document.getElementById(cell)) {
       document.getElementById(cell).value = value;
-    }
+      var e = new MouseEvent('click', {
+       'view': window,
```
Comment (david-yz-liu) [inline app/views/grade_entry_forms/_react_grades_table.js.jsx.erb]: Indent by 2 spaces

File: app/views/grade_entry_forms/_react_grades_table.js.jsx.erb [hunk 6]
```diff
@@ -16,24 +16,66 @@
...
+      }
   }
 
+  var GradeBoxTotal = React.createClass({
+    handleChange: function() {
+      this.props.value = parseInt(event.target.value);
+    },
+    render: function() {
+      return(
+        <input id={this.props.grade_id} type="text" size="4" className="grade-input" value={this.props.value} onClick={this.handleChange} disabled/>
```
Comment (david-yz-liu) [inline app/views/grade_entry_forms/_react_grades_table.js.jsx.erb]: Multiple lines here

File: app/views/grade_entry_forms/_react_grades_table.js.jsx.erb [hunk 7]
```diff
@@ -16,24 +16,66 @@
...
+    }
+  });
+
+  var GradeEntryBox = React.createClass({
+    componentDidMount: function(){
+        this.props.value = this.props.default_value;
+    },
+    handleChange: function(event) {
+      var params = {
+              'updated_grade':       parseInt(event.target.value),
```
Comment (david-yz-liu) [inline app/views/grade_entry_forms/_react_grades_table.js.jsx.erb]: Use `parseInt(_, 10)` here

File: app/views/grade_entry_forms/_react_grades_table.js.jsx.erb [hunk 8]
```diff
@@ -16,24 +16,66 @@
...
+              'student_id':          this.props.student_id,
+              'grade_entry_item_id': this.props.grade_entry_column,
+              'authenticity_token':  AUTH_TOKEN
+            };
+            jQuery.ajax({
```
Comment (david-yz-liu) [inline app/views/grade_entry_forms/_react_grades_table.js.jsx.erb]: Indentation is off

File: app/views/grade_entry_forms/_react_grades_table.js.jsx.erb [hunk 9]
```diff
@@ -16,24 +16,66 @@
...
+              url:  this.props.data_action,
+              data: params,
+              type: 'POST',
+              dataType: 'script',
+            });
+      this.props.value = parseInt(event.target.value);
+    },
+    render: function() {
+      return(
+        <input id={this.props.grade_id} type="text" size="4" className="grade-input" defaultValue={this.props.default_value} onChange={this.handleChange}/>
```
Comment (david-yz-liu) [inline app/views/grade_entry_forms/_react_grades_table.js.jsx.erb]: Break this one, too.

File: app/assets/javascripts/ReactComponents/table.js (hunk 4 above)
Comment (peterzhao333) [inline app/assets/javascripts/ReactComponents/table.js]: Not sure exactly when this happens, this was in the original code already.

**Comment by david-yz-liu** [comment]
@peterzhao333 Good work, although you should check out the changes I made, as what you did wasn't quite correct. One comment to fix: the "total" column should be read-only.

The next step should be to merge the master branch into this branch, and in particular integrate the changes you made for the bonus columns.

**Pull request merged**